Signal/src/util/OWSWindowManager.m

```diff
@@ -0,0 +1,492 @@
+    window.hidden = YES;
+    window.windowLevel = UIWindowLevel_CallView();
+    window.opaque = YES;
+    // TODO: What's the right color to use here?
+    window.backgroundColor = [UIColor ows_materialBlueColor];
+
+    UIViewController *viewController = [OWSWindowRootViewController new];
+    viewController.view.backgroundColor = [UIColor ows_materialBlueColor];
+
+    UINavigationController *navigationController =
+        [[UINavigationController alloc] initWithRootViewController:viewController];
+    navigationController.navigationBarHidden = YES;
+    self.callNavigationController = navigationController;
+
+    window.rootViewController = navigationController;
+
+    return window;
+}
+
+- (void)setIsScreenBlockActive:(BOOL)isScreenBlockActive
+{
+    OWSAssertIsOnMainThread();
+
+    _isScreenBlockActive = isScreenBlockActive;
+
+    [self ensureWindowState];
+}
+
+#pragma mark - Calls
+
+- (void)startCall:(UIViewController *)callViewController
+{
+    OWSAssertIsOnMainThread();
+    OWSAssert(callViewController);
+    OWSAssert(!self.callViewController);
+
+    self.callViewController = callViewController;
+    // Attach callViewController to window.
+    [self.callNavigationController popToRootViewControllerAnimated:NO];
+    [self.callNavigationController pushViewController:callViewController animated:NO];
+    self.isCallViewActive = YES;
+
+    [self ensureWindowState];
+}
+
+- (void)endCall:(UIViewController *)callViewController
+{
+    OWSAssertIsOnMainThread();
+    OWSAssert(callViewController);
+    OWSAssert(self.callViewController);
+
+    if (self.callViewController != callViewController) {
+        DDLogWarn(@"%@ Ignoring end call request from obsolete call view controller.", self.logTag);
+        return;
+    }
+
+    // Dettach callViewController from window.
+    [self.callNavigationController popToRootViewControllerAnimated:NO];
+    self.callViewController = nil;
+    self.isCallViewActive = NO;
+
+    [self ensureWindowState];
+}
+
+- (void)leaveCallView
+{
+    OWSAssertIsOnMainThread();
+    OWSAssert(self.callViewController);
+    OWSAssert(self.isCallViewActive);
+
+    self.isCallViewActive = NO;
+
+    [self ensureWindowState];
+}
+
+- (void)returnToCallView
+{
+    OWSAssertIsOnMainThread();
+    OWSAssert(self.callViewController);
+    OWSAssert(!self.isCallViewActive);
+
+    self.isCallViewActive = YES;
+
+    [self ensureWindowState];
+}
+
+- (BOOL)hasCall
+{
+    OWSAssertIsOnMainThread();
+
+    return self.callViewController != nil;
+}
+
+#pragma mark - Window State
+
+- (void)ensureWindowState
+{
+    OWSAssertIsOnMainThread();
+    OWSAssert(self.rootWindow);
+    OWSAssert(self.returnToCallWindow);
+    OWSAssert(self.callViewWindow);
+    OWSAssert(self.screenBlockingWindow);
+
+    if (self.isScreenBlockActive) {
+        // Show Screen Block.
+
+        [self hideRootWindowIfNecessary];
+        [self hideReturnToCallWindowIfNecessary];
```

Comment:
It's a little more complicated than that.  To avoid bad frames, we never want to hide the blocking window, so we manipulate its window level to "hide" it behind other windows.  But the other windows have fixed window level and are shown/hidden as necessary.

I've added more comments.